AssumeUTXO follow-ups

fjahr opened the pull request:
Addressing what I consider to be non- or not-too-controversial comments from #27596.

Let me know if I missed anything among the many comments that can be easily included here.

fjahr committed:
validation: remove unused mempool param in DetectSnapshotChainstate

fjahr committed:
doc: Add and edit some comments around assumeutxo

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

fjahr committed:
test: Improvements of feature_assumeutxo

- Remove usage of the internal wait_until_helper function
- Use framework self.no_op instead of new no_sync function

co-authored-by: Andrew Chow <github@achow101.com>

fjahr committed:
doc: Add snapshot chainstate removal warning to reindexing documentation

fjahr committed:
validation, test: Improve and document nChainTx check for testability

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

jamesob commented on src/node/blockstorage.cpp:
```diff
@@ -756,9 +756,10 @@ bool BlockManager::FlushChainstateBlockFile(int tip_height)
-        // blocks have been downloaded.
         return FlushBlockFile(cursor->file_num, /*fFinalize=*/false, /*finalize_undo=*/false);
     }
     return false;
```
Didn't @ryanofsky say it might be good to `return true;` here in order to avoid spurious warnings?

fjahr replied:
Right, I saw it but then missed to addressed it. Now added.

jamesob reviewed (approved):
Github ACK https://github.com/bitcoin/bitcoin/pull/28562/commits/26c815b60c8c4aec27de3fc6c7849ea3c7bdd91e

Thanks for addressing these!

fjahr committed:
blockstorage: Let FlushChainstateBlockFile return true in case of missing cursor

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

maflcko reviewed (commented):
Could also rename `HaveTxsDownloaded` to `HaveNumChainTxs` to fix the TODO in the function?

fjahr committed:
chain: Rename HaveTxsDownloaded to HaveNumChainTxs

Co-authored-by: MarcoFalke <falke.marco@gmail.com>

fjahr committed:
test: Rename wait_until_helper to wait_until_helper_internal

Co-authored-by: MarcoFalke <falke.marco@gmail.com>

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -156,19 +156,19 @@ def no_sync():
         self.log.info("Ensuring background validation completes")
         # N.B.: the `snapshot` key disappears once the background validation is complete.
-        wait_until_helper(lambda: not n1.getchainstates().get('snapshot'))
+        wait_until_helper(lambda: not n1.getchainstates().get('snapshot'), timeout_factor=2)
```
all done

fjahr commented:
Addressed feedback from @MarcoFalke

ryanofsky commented on src/init.cpp:
```diff
@@ -462,8 +462,8 @@ void SetupServerArgs(ArgsManager& argsman)
             "(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, >=%u = automatically prune block files to stay under the specified target size in MiB)", MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-reindex", "Rebuild chain state and block index from the blk*.dat files on disk. This will also rebuild active optional indexes.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-reindex-chainstate", "Rebuild chain state from the currently indexed blocks. When in pruning mode or if blocks on disk might be corrupted, use full -reindex instead.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-reindex", "Rebuild chain state and block index from the blk*.dat files on disk. This will also rebuild active optional indexes. Warning: Not yet validated snapshot chainstates will be removed.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-reindex-chainstate", "Rebuild chain state from the currently indexed blocks. When in pruning mode or if blocks on disk might be corrupted, use full -reindex instead. Warning: Not yet validated snapshot chainstates will be removed.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
In commit "doc: Add snapshot chainstate removal warning to reindexing documentation" (2bd1d39491f8ad811dcb6d3787e1e18ca0b928ad)

These warnings seem too alarmist, and also not very actionable. The point of reindex commands is to wipe state and rebuild it. So it should not be surprising that state loaded from an assumeutxo snapshot is also wiped. I think it would be a better to add a clarification instead of a warning here, and a suggestion to reload the snapshot. Maybe: "If an assumeutxo snapshot was loaded and not yet validated, it will need to be reloaded for its chain state to be rebuilt."

fjahr replied:
Done, adapted with your text.

ryanofsky reviewed (approved):
Code review ACK 69fbb77cff92649c51d5d22e06b3853db928b8e1

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2855,6 +2855,9 @@ return RPCHelpMan{
         return data;
     };
 
+    // We present the IBD/background chainstate as "normal" to users, since
+    // it should not matter to them if there is also a snapshot chainstate.
+    // To users, the IBD chainstate should always be considered "normal".
```
I think we have a simple misunderstanding of my wording, probably my English makes the comment harder to parse than it needs to be :)

My goal is to explain to developers why we talk about a "normal" chainstate when speaking to the user (via RPC interface) while internally referring to it as ibd/background in the code. I am trying to convey the same thing you said in your comment here: https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1342035181

Particularly when you say `The idea is for the "normal" chainstate to refer to a typical fully validated chainstate [...] I think the term "IBD chainstate" is basically meaningless because both chainstates are downloaded during IBD.` I am trying to say the same thing here: The normal chainstate is the ibd chainstate, no matter if there is a snapshot chainstate or not. Hence "it does not matter" that there is a snapshot with regards to the naming of the ibd/background/normal chainstate from the user perspective.

I am happy to take suggestions for better wording but I don't want to just explain what the code does. I want to document the internal/external divergence in the naming of the chainstates for future developers.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/28562#discussion_r1344824078

I implemented the changes I suggested above in #28590, so will mark this thread resolved to avoid unnecessarily slowing down this PR.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -156,19 +156,19 @@ def no_sync():
 
         self.log.info(f"Ensuring snapshot chain syncs to tip. ({FINAL_HEIGHT})")
-        wait_until_helper(lambda: n1.getchainstates()['snapshot']['blocks'] == FINAL_HEIGHT)
+        self.wait_until(lambda: n1.getchainstates()['snapshot']['blocks'] == FINAL_HEIGHT, timeout=120)
```
@MarcoFalke Did you see @achow101 's comment: https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1343167013 ?

fjahr replied:
@ryanofsky I think that's an interesting idea to set a test-wide default for the `wait_until` helper, I implemented this [here](https://github.com/bitcoin/bitcoin/pull/28562/commits/bd5e9c5cd84ffd0803bf9b8b35ec15c7afe6554f) for now but it may be better to pull it out into its own PR... What do you think about this idea @MarcoFalke ?

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -81,7 +81,7 @@ def run_test(self):
 
         def no_sync():
-            pass
+            self.no_op
```
It was actually better to remove no_sync completely and just use `self.no_op` directly. I think it's clearer then because we don't need the same no_op function declaration that we already have in the framework.

fjahr commented:
The handling of timeouts in `feature_assumeutxo` was moved out into #28586

fanquake commented:
Can you point back to / comment on #27596 in regards to what is fixed / followed up on here, and what still needs fixing? Given there are now a number of post-merge review comments.

ryanofsky reviewed (approved):
Code review ACK 2c9a84136bd4e9bdf7c85296e56364dc6a22f6b2

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/28562#issuecomment-1747088460

> Can you point back to / comment on #27596 in regards to what is fixed / followed up on here, and what still needs fixing? Given there are now a number of post-merge review comments.

This would be great, but there are a lot of open review comments on #27596 right now and this PR only addresses a small number of them, so it would be hard to compile a list of what is left and decide what is worth spending time to fix.

It would be good to comment on #27596 about what things were addressed here, though, so if someone wants to make another PR that addresses other important issues that seem important, they can do that after this one.

Sjors commented:
Can you give this a rebase? I think 952f03704958ff3f924ddf140fc9eb7663eea118 can be dropped after #28590?

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2855,6 +2855,9 @@ return RPCHelpMan{
 
+    // We present the IBD/background chainstate as "normal" to users, since
+    // it should not matter to them if there is also a snapshot chainstate.
+    // To users, the IBD chainstate should always be considered "normal".
```
21ad0c53760cf051b4daa177885db1358785d17a: this comment can be dropped now afaik since #28590 was merged.

fjahr replied:
yepp, done

Sjors reviewed (commented):
utACK 0b2c93bc560cfb0f6cbf2151bb033d67f2648891 modulo dropping commit 952f03704958ff3f924ddf140fc9eb7663eea118 and its related comment.

fjahr committed:
doc: Drop references to assumevalid in assumeutxo docs

ryanofsky commented on src/init.cpp:
```diff
@@ -462,8 +462,8 @@ void SetupServerArgs(ArgsManager& argsman)
             "(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, >=%u = automatically prune block files to stay under the specified target size in MiB)", MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-reindex", "Rebuild chain state and block index from the blk*.dat files on disk. This will also rebuild active optional indexes.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-reindex-chainstate", "Rebuild chain state from the currently indexed blocks. When in pruning mode or if blocks on disk might be corrupted, use full -reindex instead.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-reindex", "Rebuild chain state and block index from the blk*.dat files on disk. This will also rebuild active optional indexes. If an assumeutxo snapshot was loaded and not yet validated, it will need to be reloaded for its chain state to be rebuilt.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
re: https://github.com/bitcoin/bitcoin/pull/28562#discussion_r1348801054

> [bdee0fd](https://github.com/bitcoin/bitcoin/commit/bdee0fd4de85c533b799db853c514297cdc22b7c): what does this look like in practice? You start with `-reindex` and then call the RPC to load the snapshot (in time)?

Right, if you want a snapshot to be used after reindexing you have to reload the snapshot. Reindexing wipes and rebuilds chain states, and without a snapshot there is no snapshot chainstate to rebuild.

Maybe the following would be clearer: "If enabled, wipe chain state and block index, and rebuild them from the blk*.dat files on disk. Also wipe and rebuild other optional indexes that are active. If an assumeutxo snapshot was loaded and not yet validated, its chainstate will be wiped as well. The snapshot can then be reloaded via RPC."

fjahr replied:
Sounds good to me, I am using this and an adapted version for `reindex-chainste` now.

ryanofsky reviewed (approved):
Code review ACK 62db927ab516a91660de0bbf2ca7dc2c8c07ceea. Just new commit removing assumevalid mentions and rebase since last review

fjahr committed:
rpc: Use Ensure(Any)Chainman in assumeutxo related RPCs

fjahr commented:
Rebased, addressed a few more comments, dropped those that have been addressed elsewhere in the meantime. I also tried to comment on everything that is addressed here or elsewhere in #27596.

ryanofsky reviewed (approved):
Code review ACK 5d227a68627614efa8618d360efee22a47afa88b. Just suggested doc change and new EnsureChainman RPC cleanup commit since last review.

fjahr commented on src/validation.cpp:
```diff
@@ -4844,10 +4844,14 @@ void ChainstateManager::CheckBlockIndex()
+        // Make sure nChainTx sum is correctly computed.
+        unsigned int prev_chain_tx = pindex->pprev ? pindex->pprev->nChainTx : 0;
+        assert((pindex->nChainTx == pindex->nTx + prev_chain_tx)
+               // For testing, allow transaction counts to be completely unset.
```
The tests that are changed in the same commit.

fjahr replied:
> I removed the "testing" check and ran the main binary (not the tests) and it crashed here.

Hm, can you say what exactly you removed and what you can to see the crash? I couldn't reproduce that so far.

> Generally I don't think a good place to put test-only code is the main production code. Especially when it comes to consensus critical code. This makes it impossible to properly test and review the code in a production environment outside of unit tests.

I guess we should improve our test setup to include realistic `nTx` and `nChainTx` values so we can remove those checks again.
